src/librbd/io/SimpleSchedulerObjectDispatch.cc

```diff
@@ -0,0 +1,359 @@
+      return false;
+    }
+  } else {
+    delayed_requests.dispatch_time = ceph_clock_now();
+    delayed_requests.dispatch_time += utime_t(0, m_max_delay * 1000000);
+    delayed_requests.offset = object_off;
+    delayed_requests.snapc = snapc;
+    delayed_requests.op_flags = op_flags;
+
+    m_dispatch_queue.push_back(&delayed_requests);
+    if (m_dispatch_queue.front() == &delayed_requests) {
+      rescheduler_dispatch_delayed_requests();
+    }
+  }
+
+  delayed_requests.data.append(data);
+  delayed_requests.object_dispatch_flags |= object_dispatch_flags;
+  delayed_requests.requests.push_back(on_dispatched);
+
+  ldout(cct, 20) << dendl;
+
+  return true;
+}
+
+template <typename I>
+void SimpleSchedulerObjectDispatch<I>::dispatch_delayed_requests(
+    Context** on_finish) {
+  ceph_assert(m_lock.is_locked());
+  auto cct = m_image_ctx->cct;
+  ldout(cct, 20) << dendl;
+
+  for (auto &it : m_delayed_requests) {
+    dispatch_delayed_requests(it.first, on_finish);
+  }
+}
+
+template <typename I>
+void SimpleSchedulerObjectDispatch<I>::dispatch_delayed_requests(
+    uint64_t object_no, Context** on_finish) {
+  ceph_assert(m_lock.is_locked());
+
+  dispatch_delayed_requests(object_no);
+
+  auto res = m_delayed_requests.insert({object_no, {object_no}});
+  ceph_assert(res.second);
+  auto it = res.first;
+
+  auto dispatch_seq = ++m_dispatch_seq;
+  it->second.dispatch_seq = dispatch_seq;
+  auto ctx = *on_finish;
+  *on_finish = new FunctionContext(
+    [this, object_no, dispatch_seq, ctx](int r) {
+      ctx->complete(r);
+
+      Mutex::Locker locker(m_lock);
+      auto it = m_delayed_requests.find(object_no);
+      if (it == m_delayed_requests.end() ||
+          it->second.dispatch_seq != dispatch_seq) {
+        ldout(m_image_ctx->cct, 20) << "already dispatched" << dendl;
+        return;
+      }
+      dispatch_delayed_requests(it->second);
+      m_delayed_requests.erase(it);
+    });
+}
+
+template <typename I>
+void SimpleSchedulerObjectDispatch<I>::dispatch_delayed_requests(
+    uint64_t object_no) {
+  ceph_assert(m_lock.is_locked());
+  auto cct = m_image_ctx->cct;
+
+  auto it = m_delayed_requests.find(object_no);
+  if (it == m_delayed_requests.end()) {
+    ldout(cct, 20) << "object_no=" << object_no << ": not found" << dendl;
+    return;
+  }
+
+  dispatch_delayed_requests(it->second);
+  m_delayed_requests.erase(it);
+}
+
+template <typename I>
+void SimpleSchedulerObjectDispatch<I>::dispatch_delayed_requests(
+    ObjectDelayedRequests &delayed_requests) {
+  ceph_assert(m_lock.is_locked());
+  auto cct = m_image_ctx->cct;
+
+  ldout(cct, 20) << "object_no=" << delayed_requests.object_no << ", "
+                 << delayed_requests.requests.size() << " requests" << dendl;
+
+  if (delayed_requests.requests.empty()) {
+    return;
+  }
+
+  auto requests = std::move(delayed_requests.requests);
+  auto ctx = new FunctionContext(
+    [this, requests](int r) {
```

Comment:
Nit: since this just copying ```requests```, you can use C++14 generalized lambda capture (```[this, requests = std::move(delayed_requests.requests)]```